flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointOperationHandlerImpl.java

```diff
@@ -236,10 +253,17 @@ public State onExecutionsFinished() {
         @Override
         public State onAnyExecutionNotFinished(
                 Iterable<ExecutionState> notFinishedExecutionStates) {
-            terminateExceptionallyWithGlobalFailover(
+            terminateExceptionallyWithGlobalFailoverOnUnfinishedExecution(
                     notFinishedExecutionStates, completedSavepoint.getExternalPointer());
             return new FinalState();
         }
+
+        @Override
+        public State onAbort(Throwable cause) {
+            terminateExceptionallyWithGlobalFailover(
+                    cause, completedSavepoint.getExternalPointer());
+            return new FinalState();
+        }
```

Comment:
In the context of the `StopWithSavepointOperationHandlerImpl`, this method might be correct. However, in the context of the `StopWithSavepoint` state one should use it in the `onLeave` method because it can trigger a global failover. This can only work if the state is directly responsible for handling the global failover w/o going through the `AdaptiveScheduler`. That way the `StopWithSavepoint` state could say to ignore this signal if it is about to change state.

Alternatively, a method which only fails the `result` could work. However, this would add a method which is very specific to the `StopWithSavepoint` state.

A third option would be to not directly use the `StopWithSavepointOperationHandlerImpl` directly in the `StopWithSavepoint` state if it does not fit well (at least for the case of stopping the operation and not triggering a global failure and for handling the case of a failed savepoint). Only because you have hammer does not make everything a nail.